File: src/Wc3FrameExtLua/Fdf/Backdrop.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')
local Log = Wc3Utils.Log or error('')

---@type FdfDataClass
local FdfData = require('FdfData') or error('')

--=======
-- Class
--=======

local FdfBackdrop = Class.new('FdfBackdrop', FdfData)
---@class FdfBackdrop : FdfData
local public = FdfBackdrop.public
---@class FdfBackdropClass : FdfDataClass
local static = FdfBackdrop.static
---@type FdfBackdropClass
local override = FdfBackdrop.override
local private = {}

--=========
-- Static
--=========

---@param name string
---@param child FdfBackdrop | nil
---@return FdfBackdrop
function override.new(name, child)
    isTypeErr(name, 'string', 'name')
    if child then isTypeErr(child, FdfBackdrop, 'child') end

    local instance = child or Class.allocate(FdfBackdrop)
    instance = FdfData.new(name, 'BACKDROP', false, nil, instance)

    return instance
end

--========
-- Public
--========

---@param width number
function public:setWidth(width)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(width, 'number', 'width')

    local str_width = tostring(width)
    self:setParameter('Width', str_width)
end

---@param height number
function public:setHeight(height)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(height, 'number', 'height')

    local str_height = tostring(height)
    self:setParameter('Height', str_height)
end

--- Look-Up FileNames in some String table (for example gameinterface)
---@param flag boolean
function public:setDecorateFileNames(flag)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(flag, 'boolean', 'flag')

    if flag then
        self:setParameter('DecorateFileNames')
    else
        self:removeParameter('DecorateFileNames')
    end
end

---@param flag boolean
function public:setAllPoints(flag)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(flag, 'boolean', 'flag')

    if flag then
        self:setParameter('SetAllPoints')
    else
        self:removeParameter('SetAllPoints')
    end
end

--- Fills the Frame with tiles of the texture. Without the background file is stretched
---@param flag boolean
function public:setBackgroundTileMode(flag)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(flag, 'boolean', 'flag')

    if flag then
        self:setParameter('BackdropTileBackground')
    else
        self:removeParameter('BackdropTileBackground')
    end
end

--- In Tile mode, the size of each tile. [0, 1]
---@param size number
function public:setBackgroundTileSize(size)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(size, 'number', 'size')

    if not self:getParameter('BackdropTileBackground') then
        Log:wrn('Parameter \"BackdropBackgroundSize\" works in tile mode only')
    end

    size = size > 1 and 1 or size < 0 and 0 or size
    self:setParameter('BackdropBackgroundSize', tostring(size))
end

--- Main image
---@param path string
function public:setBackground(path)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(path, 'string', 'path')

    self:setParameter('BackdropBackground', '\"'..path..'\"')
end

--- Cuts in the background from the frames outside, granting space between
--- the edge and the background. Right Top Bottom Left. The numbers can be
--- Negative to extend this side over the edge into the outside.
---@param right number
---@param top number
---@param bottom number
---@param left number
function public:setInsets(right, top, bottom, left)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(right, 'number', 'right')
    isTypeErr(top, 'number', 'top')
    isTypeErr(bottom, 'number', 'bottom')
    isTypeErr(left, 'number', 'left')

    self:setParameter('BackdropBackgroundInsets',
                      string.format('%f %f %f %f', right, top, bottom, left))
end

--- FRAMEPOINTS having borders (edgefile). Order does not matter
--- "UL|UR|BL|BR|T|L|B|R"
---@param flags string
function public:setCornerFlags(flags)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(flags, 'string', 'flags')

    self:setParameter('BackdropCornerFlags', '\"'..flags..'\"')
end

--- Size of the border/edge.
---@param size number
function public:setCornerSize(size)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(size, 'number', 'size')

    size = size > 1 and 1 or size < 0 and 0 or size
    self:setParameter('BackdropCornerSize', tostring(size))
end

--- The border File, this texture having the border as fragments
--- next to each other (gameinterface). You don't need the single
--- ones if you have this one.
---@param path string
function public:setEdgeFile(path)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(path, 'string', 'path')

    self:setParameter('BackdropEdgeFile', '\"'..path..'\"')
end

---@param path string
function public:setCornerEdgeFile(path)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(path, 'string', 'path')

    self:setParameter('BackdropCornerFile', '\"'..path..'\"')
end

---@param path string
function public:setLeftEdgeFile(path)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(path, 'string', 'path')

    self:setParameter('BackdropLeftFile', '\"'..path..'\"')
end

---@param path string
function public:setTopEdgeFile(path)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(path, 'string', 'path')

    self:setParameter('BackdropTopFile', '\"'..path..'\"')
end

---@param path string
function public:setRightEdgeFile(path)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(path, 'string', 'path')

    self:setParameter('BackdropRightFile', '\"'..path..'\"')
end

---@param path string
function public:setBottomEdgeFile(path)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(path, 'string', 'path')

    self:setParameter('BackdropBottomFile', '\"'..path..'\"')
end

--- Use transparency by alpha channels. An image with alpha channel
--- transparency but without this will look wierd and wrong. It also
--- allows to see the things behind the frame.
---@param flag boolean
function public:setBlendAll(flag)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(flag, 'boolean', 'flag')

    if flag then
        self:setParameter('BackdropBlendAll')
    else
        self:removeParameter('BackdropBlendAll')
    end
end

--- Vertical, a face looking to the left will look with this option enabled to the right.
---@param flag boolean
function public:setMirrired(flag)
    isTypeErr(self, FdfBackdrop, 'self')
    isTypeErr(flag, 'boolean', 'flag')
    
    if flag then
        self:setParameter('BackdropMirrored')
    else
        self:removeParameter('BackdropMirrored')
    end
end

--=========
-- Private
--=========

return static

File: src/FrameExt/Fdf/Backdrop.ts
import { Fdf } from "../Fdf";

type BackdropCornerFlags = 'UL'|'UR'|'BL'|'BR'|'T'|'L'|'B'|'R'

export class FdfBackdrop extends Fdf {
    constructor(name: string){
        super(name, 'BACKDROP', false)
    }

    public get width(){return this._width}
    public set width(w: number){
        this._setParam('Width', w.toString())
        this._width = w
    }

    public get height(){return this._height}
    public set height(h: number){
        this._setParam('Height', h.toString())
        this._height = h
    }

    public get decorateFileNames(){return this._decorate}
    public set decorateFileNames(flag: boolean){
        if (flag){
            this._setParam('DecorateFileNames')
        } else {
            this._removeParam('DecorateFileNames')
        }
        this._decorate = flag
    }

    public get setAllPoints(){return this._all_points}
    public set setAllPoints(flag: boolean){
        if (flag){
            this._setParam('SetAllPoints')
        } else {
            this._removeParam('SetAllPoints')
        }
        this._all_points = flag
    }

    public get backgroundTileMode(){return this._tile_mode}
    public set backgroundTileMode(flag: boolean){
        if (flag){
            this._setParam('BackdropTileBackground')
        } else {
            this._removeParam('BackdropTileBackground')
        }
        this._tile_mode = flag
    }

    public get backgroudTileSize(){return this._tile_size}
    public set backgroudTileSize(size: number){
        this._setParam('BackdropBackgroundSize', size.toString())
        this._tile_size = size
    }

    public get background(){return this._background}
    public set background(path: string){
        this._setParam('BackdropBackground', path)
        this._background = path
    }

    public get insets(){return this._insets}
    public set insets(insets: [right: number, top: number, bottom: number, left: number]){
        let [right, top, bottom, left] = insets
        this._setParam('BackdropBackgroundInsets',
                        string.format('%f %f %f %f', right, top, bottom, left))
        this._insets = insets
    }

    public get cornerFlags(){return this._cornerFlags}
    public set cornarFlags(flags: BackdropCornerFlags[]){
        let s_flags = flags.length > 0 ? flags[0] : ''
        for (let i = 1; i < flags.length; i++){
            s_flags += '|' + flags[i]
        }
        this._setParam('BackdropCornerFlags', '\"' + s_flags + '\"')
        this._cornerFlags = flags
    }

    public get cornerSize(){return this._cornerSize}
    public set cornerSize(size: number){
        size = size > 1 ? 1 : size < 0 ? 0 : size
        this._setParam('BackdropCornerSize', size.toString())
        this._cornerSize = size
    }

    public get edgeFile(){return this._edgeFile}
    public set edgeFile(path: string){
        this._setParam('BackdropEdgeFile', '\"' + path + '\"')
        this._edgeFile = path
    }
    
    public get blendAll(){return this._blend}
    public set blendAll(flag: boolean){
        if (flag){
            this._setParam('BackdropBlendAll')
        } else {
            this._removeParam('BackdropBlendAll')
        }
        this._blend = flag
    }

    public get mirrored(){return this._mirrored}
    public set mirrored(flag: boolean){
        if (flag){
            this._setParam('BackdropMirrored')
        } else {
            this._removeParam('BackdropMirrored')
        }
        this._mirrored = flag
    }

    private _width: number = -1;
    private _height: number = -1;
    private _decorate: boolean = false;
    private _all_points: boolean = false;
    private _tile_mode: boolean = false;
    private _tile_size: number = 1;
    private _background: string = '';
    private _insets: [number, number, number, number] = [0, 0, 0, 0];
    private _cornerFlags: BackdropCornerFlags[] = [];
    private _cornerSize: number = 0;
    private _edgeFile: string = '';
    private _blend: boolean = false;
    private _mirrored: boolean = false;
}